### PR TITLE
sdel fixes and a whole load of other shit

### LINKED ISSUE #65: We need to use garbageCleanup() for a buttload more objects / High GC Failure rate
Example Sdel controller stat outputs:
```Garbage | 0ds | (CPU:0%) | (Left: 128) | (Failed: 84.7981%)```
```Garbage | 0ds | (CPU:0%) | (Left: 7) | (Failed: 99.1755%)```

The game ends up with around 85-99% Failure (that is to say, forced Deletion) of each datum due to garbageCleanup() not being used to remove references (to be fair I only just implemented the changes to Sdel and added garbageCleanup() today 13/06/2015)

Should be a simple fix, just define proper working garbageCleanup()s for each type we have that has lots of references.

Amusingly, I have never seen it reach actual 100%, only 99% so SOMETHING out there is garbage collecting correctly!

### PR DESCRIPTION
**Pull request by DaedalusGame**
Look if I knew I would write a changelog here.

fixes #65 almost definitely.

### CODE REVIEW

File: code/world/subsystems/sdel.dm [hunk 1]
```diff
@@ -56,14 +54,29 @@ datum/controller/sdel/getStat()
...
-			del(D)
-
-	scheck()
-
+	for(var/dref in deleted)
+		var/datum/D = locate(dref)
+		var/gctime = deleted[dref]
+
+		if(!D)
+			deleted.Remove(dref)
```
Comment (RemieRichards) [inline code/world/subsystems/sdel.dm]: please don't use ```Remove()```, ```-=``` is so much better

File: code/world/subsystems/sdel.dm [hunk 2]
```diff
@@ -56,14 +54,29 @@ datum/controller/sdel/getStat()
...
+				failedSDeletions++
+				failedTypes["[D.type]"]++
+				del(D)
+			deleted.Remove(dref)
+
+	scheck() //why is this out here!?
+
+//This is a debug verb
+/client/verb/getFailedTypeDeletes()
+	set name = "why is there so many failures!?"
```
Comment (RemieRichards) [inline code/world/subsystems/sdel.dm]: Why ARE there so many failures

File: code/world/subsystems/sdel.dm [hunk 3]
```diff
@@ -56,14 +54,29 @@ datum/controller/sdel/getStat()
...
+		if(!D)
+			deleted.Remove(dref)
+			continue
+		else if((D.garbageTime + GARBAGE_PATIENCE) < world.time)
+			if(D.garbageTime && D.garbageTime == gctime) //Only hard-delete if necessary and it's the same object. I hope this works.
+				deleted -= D
+				failedSDeletions++
+				failedTypes["[D.type]"]++
+				del(D)
+			deleted.Remove(dref)
```
Comment (RemieRichards) [inline code/world/subsystems/sdel.dm]: see above

**Comment by RemieRichards** [comment]
Besides my comments, this is good, fix those though please @DaedalusGame

File: code/world/subsystems/sdel.dm (hunk 1 above)
Comment (DaedalusGame) [inline code/world/subsystems/sdel.dm]: observe that -= doesn't work on associative lists bby

Comment (RemieRichards) [inline code/world/subsystems/sdel.dm]: tell that to /tg/
```queue -= "\ref[A]" ```

Comment (DaedalusGame) [inline code/world/subsystems/sdel.dm]: tg a shit

File: code/mob/player/player.dm [hunk 4]
```diff
@@ -44,6 +44,10 @@
 		sdel(D)
 	playerEquipped = null
 
+	for(var/datum/O in playerSpellHolders)
```
Comment (RemieRichards) [inline code/mob/player/player.dm]: this already exists in the ```garbageCleanup()``` a few lines down.

**Comment by RemieRichards** [comment]
:P

File: code/world/subsystems/sdel.dm (hunk 1 above)
Comment (DaedalusGame) [inline code/world/subsystems/sdel.dm]: pretty sure i tested it before and it didn't work but now it does. Odd.

Comment (RemieRichards) [inline code/world/subsystems/sdel.dm]: @DaedalusGame it's called ~Byond~

File: code/mob/player/player.dm (hunk 4 above)
Comment (DaedalusGame) [inline code/mob/player/player.dm]: i can't believe bord's fucking retarded

Comment (RemieRichards) [inline code/mob/player/player.dm]: @DaedalusGame you're aware you removed the wrong one, right? haha

Comment (DaedalusGame) [inline code/mob/player/player.dm]: they are both the same thing tho ;3;

no bully

Comment (RemieRichards) [inline code/mob/player/player.dm]: @DaedalusGame no I mean you literally remove the wrong one!
You didn't remove my copy of playerSpellHolders, you removed the one about statuseffects! haha

**Comment by sawu-tg** [comment]
@RemieRichards stop bullying bord so much and come on IRC

**Comment by RemieRichards** [comment]
@sawu-tg I can't IRC at college ;_;

**Comment by sawu-tg** [comment]
@RemieRichards then go proof-read the world story on the wiki

**Comment by DaedalusGame** [comment]
Did I fix everything ;3;

**Comment by RemieRichards** [comment]
@DaedalusGame besides the enrgish on the debug button ya :P

**Comment by DaedalusGame** [comment]
>enrgish

I AIN'T HAVIN ANY OF DAT SHIT

**Pull request merged**